File: js/src/instructions/createInstruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInstruction = void 0;
const buffer_1 = require("buffer");
const web3_js_1 = require("@solana/web3.js");
const int_1 = require("../int");
function createInstruction(nameProgramId, systemProgramId, nameKey, nameOwnerKey, payerKey, hashed_name, lamports, space, nameClassKey, nameParent, nameParentOwner) {
    const buffers = [
        buffer_1.Buffer.from(Int8Array.from([0])),
        new int_1.Numberu32(hashed_name.length).toBuffer(),
        hashed_name,
        lamports.toBuffer(),
        space.toBuffer(),
    ];
    const data = buffer_1.Buffer.concat(buffers);
    const keys = [
        {
            pubkey: systemProgramId,
            isSigner: false,
            isWritable: false,
        },
        {
            pubkey: payerKey,
            isSigner: true,
            isWritable: true,
        },
        {
            pubkey: nameKey,
            isSigner: false,
            isWritable: true,
        },
        {
            pubkey: nameOwnerKey,
            isSigner: false,
            isWritable: false,
        },
    ];
    if (nameClassKey) {
        keys.push({
            pubkey: nameClassKey,
            isSigner: true,
            isWritable: false,
        });
    }
    else {
        keys.push({
            pubkey: new web3_js_1.PublicKey(buffer_1.Buffer.alloc(32)),
            isSigner: false,
            isWritable: false,
        });
    }
    if (nameParent) {
        keys.push({
            pubkey: nameParent,
            isSigner: false,
            isWritable: false,
        });
    }
    else {
        keys.push({
            pubkey: new web3_js_1.PublicKey(buffer_1.Buffer.alloc(32)),
            isSigner: false,
            isWritable: false,
        });
    }
    if (nameParentOwner) {
        keys.push({
            pubkey: nameParentOwner,
            isSigner: true,
            isWritable: false,
        });
    }
    return new web3_js_1.TransactionInstruction({
        keys,
        programId: nameProgramId,
        data,
    });
}
exports.createInstruction = createInstruction;

File: python/src/instructions/create_instruction.py
from typing import Optional

from solders.instruction import Instruction, AccountMeta
from solders.pubkey import Pubkey

from int import Numberu64, Numberu32


def create_instruction(
    name_program_id: Pubkey,
    system_program_id: Pubkey,
    name: Pubkey,
    name_owner: Pubkey,
    payer: Pubkey,
    hashed_name: bytes,
    lamports: Numberu64,
    space: Numberu32,
    name_class: Optional[Pubkey],
    name_parent: Optional[Pubkey],
    name_parent_owner: Optional[Pubkey],
) -> Instruction:
    buffers = [
        bytes([0]),
        Numberu32(len(hashed_name)).to_bytes(),
        hashed_name,
        lamports.to_bytes(),
        space.to_bytes(),
    ]

    data = b"".join(buffers)

    accounts = [
        AccountMeta(pubkey=system_program_id, is_signer=False, is_writable=False),
        AccountMeta(pubkey=payer, is_signer=True, is_writable=True),
        AccountMeta(pubkey=name, is_signer=False, is_writable=True),
        AccountMeta(pubkey=name_owner, is_signer=False, is_writable=False),
    ]

    if name_class:
        accounts.append(
            AccountMeta(pubkey=name_class, is_signer=True, is_writable=False)
        )
    else:
        accounts.append(
            AccountMeta(pubkey=Pubkey([0] * 32), is_signer=False, is_writable=False)
        )
    if name_parent:
        accounts.append(
            AccountMeta(pubkey=name_parent, is_signer=False, is_writable=False)
        )
    else:
        accounts.append(
            AccountMeta(pubkey=Pubkey([0] * 32), is_signer=False, is_writable=False)
        )
    if name_parent_owner:
        accounts.append(
            AccountMeta(pubkey=name_parent_owner, is_signer=True, is_writable=False)
        )

    return Instruction(
        program_id=name_program_id,
        accounts=accounts,
        data=data,
    )
